### PR TITLE
[WIP] tf.Variable like API for variables

### LINKED ISSUE #170: Variable class (like tf.Variable) that supports eager mode
I've been looking at adding a tf.Variable like class, that would work in eager and graph mode (the variable op doesn't work in eager), and register itself in the execution environment.  It's not terribly useful until we get eager gradient support, but it's not dependent on it either so there's no reason to wait.

Eager mode is easy, just have a field.  For graph, we can store a Variable op, assign to it, but return/expose the return of the assign (it's the new value).  The assigns can take the last assign as a control dep.

I've got a draft [here](https://github.com/tensorflow/java/compare/master...rnett:rn_variable) I can PR, but I'm not sure how much the type system refactor would affect it.  At a minimum I don't see a reason to merge before #160.

### PR DESCRIPTION
**Pull request by rnett**
Fixes #170.

Adds a `tf.Variable` like class, using the new resource API (see [here](https://www.tensorflow.org/api_docs/python/tf/compat/v1/enable_resource_variables) and [here](https://github.com/tensorflow/tensorflow/blob/v2.4.0/tensorflow/python/ops/resource_variable_ops.py)).  It's compatible with Eager and Graph mode, and will work better with eager gradient tapes in the future.

A couple of points I'd like feedback on:
1. Immutability by default.  Most of the time variables are only updated by the optimizer, not by any code users will write, so I think this makes sense.  It's also not hard to access the mutable version (`asMutableVariable()`).
2. Auto-scoping operations to the creation scope.  This is the `value()`, `create()`, etc. methods that don't take a `Scope` parameter and use the scope the variable is created with.  While these are certainly much nicer to use, they do muddle the semantics a bit, and if gradient tapes are done via Scope it would probably cause issues (since the tapes need to be notified of variable usage, afaik).  I originally did it this way because `shape()` depends on `value()` which needs a scope, although it is possible to just return the variable's initial shape.

### CODE REVIEW

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/MutableVariable.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+    if (!hasInitialized) {
+      throw new IllegalStateException("Variable has not been initialized, can not get.");
+    }
+    if (cachedRead == null) {
+      if (lastAssign != null) {
+        scope = scope.withControlDependencies(Collections.singletonList(lastAssign));
+      }
+      cachedRead = ReadVariableOp.create(scope, handle, tType);
+    }
+    return cachedRead;
```
Comment (Craigacp) [inline tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/MutableVariable.java:115]: This method has data races. In Python they don't need to worry about it, but we should at least consider what the possible behaviours are in Java.

**Review by Craigacp — COMMENTED** [review]
I don't think splitting out mutability is a good choice here. The variable can be mutated under the covers by the TF runtime (and will be, because that's what it's for), and the documentation here isn't clear that this is the case. Also an immutable variable sounds a lot like `constant` and we've already got those, which these docs don't reference. I'm not clear what's gained by making the split given the only implementation is mutable, and the immutable view has a method that returns the mutable version. Could you give me more of an idea why having an unmodifiable view be the default is worthwhile?

I think using the reference variables is fine and we should probably migrate, but there are a bunch of things in the framework which gloss over the existence of such variables and we'll need to migrate those (as when I ported over the optimizers from Python I ignored all the reference ops and use the regular ones, but presumably we'll need to switch to using the reference ops).

**Comment by rnett** [comment]
`isValueInitialized()`, which uses `IsVariableInitialized`, requires a `ref` tensor, is there any way to get this in the Java API yet?

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/ExecutionEnvironment.java [hunk 2]
```diff
@@ -41,6 +44,15 @@
...
   Types environmentType();
 
+  Map<String, MutableVariable<?>> variables();
+
+  /**
+   * Registers a variable with this execution environment.
+   * @deprecated Done automatically in Variable's constructor, should only be used internally.
+   */
+  @Deprecated
+  void registerVariable(MutableVariable<?> variable);
```
[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/ExecutionEnvironment.java:54]: Yeah, that's correct, I used `@Deprecated` as I'm not aware of a more semantically correct version, and just noting it in the doc seemed insufficient.  Is there a more correct strategy for this?

**Comment by rnett** [comment]
The immutable (more properly read only) version is somewhat similar to Kotlin's collections.  It's always backed by a mutable version, and may be casted and mutated in most cases, but the mutation APIs are hidden.  So not it's immutable, just a read only view of the mutable variable, semantically.  I do need to update the docs a bit to reflect that.  The idea was since that generally users assigning to a variable will be an error, it should be harder to do and be made more explicit.

**Comment by rnett** [comment]
Optimizers definitely will need a pass, I originally didn't do it as part of this PR as I wasn't going to use the resource variables, but now that I am I probably should.  I'd like to make it a bit easier to pass in the list of variables to update on rather than using the entire graph, too, for multi-model support (i.e. GANs).

One other thing that came up: I can't use initialization as a control dependency since it would be re-ran each time.  Is there a way to require initialization to have been ran in the session?  You will get an uninitialized error if you don't run it, but I'd like a clearer way (or ideally to run it automatically, like a control dependency, just only once per session).

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/MutableVariable.java (hunk 1 above)
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/MutableVariable.java:115]: Yeah, I should use a local variable there.  Do you think using `AtomicReference` as well is worth it?  I'm not sure how much of the rest is thread safe.

**Comment by rnett** [comment]
Currently somewhat blocked by tensorflow/tensorflow#46114, although if/when we add gradient registry ops we could work around it.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/ExecutionEnvironment.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/ExecutionEnvironment.java:54]: Refactor it so that it's not visible, or document that it's not for external use (though that's less than ideal). The deprecated tag will probably cause the linter to fail on TF-java builds when we re-enable it, so we can't do that.

**Comment by Craigacp** [comment]
> The immutable (more properly read only) version is somewhat similar to Kotlin's collections. It's always backed by a mutable version, and may be casted and mutated in most cases, but the mutation APIs are hidden. So not it's immutable, just a read only view of the mutable variable, semantically. I do need to update the docs a bit to reflect that. The idea was since that generally users assigning to a variable will be an error, it should be harder to do and be made more explicit.

Sure, but that's not how Java's collections work and I think it might be preferable not to leak Kotlin-isms into the Java API used by all the JVM languages. Doesn't python expose these operations to allow updates to epoch numbers, learning rates and similar? I think those would be common use cases we'd have in Java too.

> Currently somewhat blocked by [tensorflow/tensorflow#46114](https://github.com/tensorflow/tensorflow/issues/46114), although if/when we add gradient registry ops we could work around it.

The lack of this gradient (which presumably just passes straight through) means it's not possible to train any models which contain this variable right?

**Comment by rnett** [comment]
> The lack of this gradient (which presumably just passes straight through) means it's not possible to train any models which contain this variable right?

Yeah, and it is just a passthrough.

> Doesn't python expose these operations to allow updates to epoch numbers, learning rates and similar? I think those would be common use cases we'd have in Java too.

Yeah, true, I'll refactor that.  If/when proper logging is added it would be good to have a warning when variables are mutated when they aren't "supposed" to be (i.e. in the forward pass).

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/ExecutionEnvironment.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/ExecutionEnvironment.java:54]: Unfortunately, it's in an interface, so javadoc is the best I can do.  Manual use shouldn't cause too many issues though, since it's keyed on a final field.

**Comment by Craigacp** [comment]
> > The lack of this gradient (which presumably just passes straight through) means it's not possible to train any models which contain this variable right?
> 
> Yeah, and it is just a passthrough.

Ok, well this needs to go on hold till they fix that upstream then.

> 
> > Doesn't python expose these operations to allow updates to epoch numbers, learning rates and similar? I think those would be common use cases we'd have in Java too.
> 
> Yeah, true, I'll refactor that. If/when proper logging is added it would be good to have a warning when variables are mutated when they aren't "supposed" to be (i.e. in the forward pass).

How would we identify what's a forward pass vs a backward pass? Privileging the target names seems restrictive.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/Variable.java [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+import org.tensorflow.op.Operands;
+import org.tensorflow.op.Scope;
+import org.tensorflow.op.annotation.Endpoint;
+import org.tensorflow.op.annotation.Operator;
+import org.tensorflow.proto.framework.DataType;
+import org.tensorflow.types.TBool;
+import org.tensorflow.types.family.TType;
+
+/**
+ * A class representing a read-only tensor variable with a constant shape and data type.  Analogous to Python's tf.Variable.
```
[the remark this answers is not in the document]
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/Variable.java:44]: The docs still say it's read only when it's not.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/Variable.java [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+ */
+@Operator
+public class Variable<T extends TType> implements Operand<T> {
+
+  private final Scope initialScope;
+  private final String name;
+
+  private final Shape shape;
+  private final DataType dataType;
+  private final boolean trainable;
```
[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/Variable.java:63]: It's not in python, so I copied that.  It doesn't seem that hard to do, but we'd want to have something to ensure it's not changed while a gradient tape is active in eager mode, or have some special handling for that.  It depends on what that API looks like though.

I don't know how keras handles it.  Maybe with the StopGradient op?

**Comment by rnett** [comment]
> Ok, well this needs to go on hold till they fix that upstream then.

Yeah.  I made a PR following the C++ gradient instructions, but it doesn't work.  I'm going to follow up in the mailing list but haven't had time yet.

> How would we identify what's a forward pass vs a backward pass? Privileging the target names seems restrictive.

Yeah, I don't know how feasible that is.  It would probably have to be a framework level setting which we shouldn't mix into core.

File: tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/Variable.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/Variable.java:63]: Ah ok, so the python docs make sense. The notes on the behaviour of trainable in saved model means it's still counter-intuitive as I think I'd like to store the epoch count in a checkpoint, it makes it easier to restart training, but oh well.

[the remark this answers is not in the document]
Comment (Craigacp) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/Variable.java:63]: It would be nice to add the trainable variables to a hook in the graph the same way they do in Python. The registerVariable hook could check if it's trainable.

[the remark this answers is not in the document]
Comment (rnett) [reply tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/variable/Variable.java:63]: I could, yeah.  I figured it was trivial to filter the variables list.  I would think any optimizer implementations would want to take a list of variables from whatever types of models they use anyways, to allow for multiple models in the same graph.

**Comment by rnett** [comment]
This will change fairly significantly due to #237 (and parts are currently broken), so don't merge it.  However, with https://github.com/tensorflow/tensorflow/pull/46115, the gradient works.

**Comment by JimClarke5** [comment]
In `Metrics` and `Model`, several variables are updated outside of Optimizers.

**Comment by JimClarke5** [comment]
`callbacks.LearningRateScheduler` requires a way to update the `learningRate` in an `Optimizer`. The original thought was to use `Placeholders` for this, but if we could use a `Variable` we wouldn't need to do funny stuff with feeds and changing `Tensor` values.
